drivers/android/transaction.rs

```diff
@@ -79,7 +106,13 @@ impl Transaction {
             offsets_size: tr.offsets_size as _,
             links: Links::new(),
             free_allocation: AtomicBool::new(true),
-        })
+        })?;
+
+        let mut_tr = Arc::get_mut(&mut tr).ok_or(Error::EINVAL)?;
+
+        // SAFETY: `inner` is pinned behind `Arc`.
+        kernel::spinlock_init!(Pin::new_unchecked(&mut_tr.inner), "Transaction::inner");
```

Comment:
I'm still trying to get my head around `Pin`ning. Is there a guarantee that `Self` will never move out of the `Arc`?
I'm guessing you either need an explicit guarantee (in the form of an `# Invariant`, might be hard to mentally enforce here) or return `Pin<Arc<Self>>` ?

Reply:
It's not guaranteed. At the moment it's an invariant of the project: I don't move objects behind `Arc`s.

Some background: in the beginning I had a bunch of `TODO: This needs to be pinned` in a lot of places, and a bunch of `Pin`s that resulted in a bunch of `unsafe` blocks (it's always unsafe to get a mutable reference to a pinned type because the compiler needs us to manually ensure that we don't move anything). I removed them until we get a better handle on how this will look like.